### ISSUE TITLE
Theo 3.0.0

### ISSUE THREAD

**Issue by niklassaers** [GraphStory/neo4j-ios#24]
Hey @kwylez, I'm hacking mostly on my own initiative on Theo, but I was wondering what from the backlog should be done for Theo to reach 3.0.0? Anything you need a hand with?

**Comment by kwylez** [comment]
I think what you submitted gets the library is a _great_ position to cut a 3.0 release. I'll look at the other two issues you asked about and see if I can get those knocked out and then tag and push the new release to `master`

**Comment by niklassaers** [comment]
Cool. I've identified a problem when running server side where the server will "hog" the main thread. The code directly uses main thread many places for callbacks - though not all, I haven't really understood which is when. But this in effect stops the callbacks server side. For now I've put it on a concurrent queue, but I should probably write a little bit to make it easy for the developer to choose callback queue, and possibly have good defaults depending on platform.

**Comment by niklassaers** [comment]
I'm a bit uncertain, for instance in Client's metaDescription, the "happy path" leads to completionBlock being run on the main thread, whereas errors, exceptions and "no data" have completionBlock running on the concurrent thread. Is this intentional? (because if it is, I'm probably about to break this having them all run on the concurrent queue)

**Comment by kwylez** [comment]
@niklassaers that is a bug. _Ideally_ there should be a property on the `Client` class that will default to the main thread by default, but allow the user to specify which thread they want callbacks to come back to.

**Comment by kwylez** [comment]
@niklassaers I'll add that in now and push it up

**Comment by niklassaers** [comment]
@kwylez Sorry for the confusion introduced, but I just submitted pull request #25 which has it compile on Linux, and takes out all the calls to the main thread. In doing that, I figured the user can, in the completion block, decide if it needs to be moved to another thread or not. While not as convenient, it saved going back and forth between threads if not necessary. You can decide what you prefer, and I'll merge the changes if you like.

**Comment by niklassaers** [comment]
Cool, I'm happy we reached the same conclusion. :-)

**Comment by niklassaers** [comment]
I'm in the process of adding a bit more test support so that the tests will also run under Linux, btw

**Comment by kwylez** [comment]
Wow! That's a pretty big enhancement

**Comment by niklassaers** [comment]
It does make some changes. To be able to do "swift test" I've had to move the Source/Theo/TheoTests to Tests/TheoTests. I've updated the xcodeproj accordingly so it runs nicely there. Hope you don't mind. I was hoping I could define it in Package.swift, but if I understood correctly that's a feature waiting to be implemented. With what's available in https://github.com/niklassaers/neo4j-ios/tree/fix/linux-tests I'm able to do "swift test" now. Tests fail, and I'm trying to figure out if I can attach my Xcode debugger in any way, instead of having to run lldb by hand.
